File: 08 August/08-28/second_best.py
def get_laptop_cost(laptops, budget):
    prices = laptops[:]
    prices.sort(reverse = True)
    prices = list(dict.fromkeys(prices))

    if len(prices) >= 2 and prices[1] <= budget:
        return prices[1]

    else:
        for i in range(2, len(prices) - 1):
            if prices[i] <= budget:
                return prices[i]

    return 0

print(get_laptop_cost([1500, 2000, 1800, 1400], 1900))
print(get_laptop_cost([1500, 2000, 2000, 1800, 1400], 1900))
print(get_laptop_cost([2099, 1599, 1899, 1499], 2200))
print(get_laptop_cost([2099, 1599, 1899, 1499], 1000))
print(get_laptop_cost([1200, 1500, 1600, 1800, 1400, 2000], 1450))
print(get_laptop_cost([1000, 2000], 500))
print(get_laptop_cost([1000, 2000], 3000))
print(get_laptop_cost([1000, 2000], 1500))
print(get_laptop_cost([1000], 1500))

# --- TEST SUITE ---

tests_text = '''
1. get_laptop_cost([1500, 2000, 1800, 1400], 1900) should return 1800
2. get_laptop_cost([1500, 2000, 2000, 1800, 1400], 1900) should return 1800
3. get_laptop_cost([2099, 1599, 1899, 1499], 2200) should return 1899
4. get_laptop_cost([2099, 1599, 1899, 1499], 1000) should return 0
5. get_laptop_cost([1200, 1500, 1600, 1800, 1400, 2000], 1450) should return 1400
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 08 August/08-28/secondBest.js
function getLaptopCost(laptops, budget) {
  const prices = [...new Set(laptops.sort((a, b) => b - a))];

  if (prices[1] <= budget) {
    return prices[1];
  }

  else {
    for (let i = 2; i < prices.length; i ++) {
      if (prices[i] <= budget)
        return prices[i];
    }
  }

  return 0;
}

console.log(getLaptopCost([1500, 2000, 1800, 1400], 1900));
console.log(getLaptopCost([1500, 2000, 2000, 1800, 1400], 1900));
console.log(getLaptopCost([2099, 1599, 1899, 1499], 2200));
console.log(getLaptopCost([2099, 1599, 1899, 1499], 1000));
console.log(getLaptopCost([1200, 1500, 1600, 1800, 1400, 2000], 1450));
console.log(getLaptopCost([1000, 2000], 500));
console.log(getLaptopCost([1000, 2000], 3000));
console.log(getLaptopCost([1000, 2000], 1500));
console.log(getLaptopCost([1000], 1500));

// --- TEST SUITE ---

const testsText = `
1. getLaptopCost([1500, 2000, 1800, 1400], 1900) should return 1800
2. getLaptopCost([1500, 2000, 2000, 1800, 1400], 1900) should return 1800
3. getLaptopCost([2099, 1599, 1899, 1499], 2200) should return 1899
4. getLaptopCost([2099, 1599, 1899, 1499], 1000) should return 0
5. getLaptopCost([1200, 1500, 1600, 1800, 1400, 2000], 1450) should return 1400
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);